base/math.jl

```diff
@@ -81,18 +81,139 @@ function clamp!(x::AbstractArray, lo, hi)
+17
+```
+"""
+evalpoly(x, p::AbstractVector) = _evalpoly(x, p)
+
+function _evalpoly(x, p)
+    N = length(p)
+    ex = p[end]
+    for i in N-1:-1:1
+        ex = muladd(x, ex, p[i])
+    end
+    ex
+end
+
+"""
+    evalpoly(z::Complex, p::Tuple)
+Evaluate the polynomial ``\\sum_k p[k] z^{k-1}`` for the coefficients `p[1]`, `p[2]`, ...;
+that is, the coefficients are given in ascending order by power of `z`. This function
+generates efficient code using a Goertzel-like algorithm specialized for complex arguments
+with loops unrolled at compile time.
+The Goertzel-like algorthim is described in Knuth's Art of Computer Programming,
+Volume 2: Seminumerical Algorithms, Sec. 4.6.4.
+# Example
+```jldoctest
+julia> evalpoly(2 + im, (1, 2, 3))
+14 + 14im
+```
+"""
+function evalpoly(z::Complex, p::Tuple)
+    if @generated
+        N = length(p.parameters)
+        a = :(p[end])
+        b = :(p[end-1])
+        as = []
+        for i in N-2:-1:1
+            ai = Symbol("a", i)
+            push!(as, :($ai = $a))
+            a = :(muladd(r, $ai, $b))
+            b = :(p[$i] - s * $ai)
+        end
+        ai = :a0
+        push!(as, :($ai = $a))
+        C = Expr(:block,
+                 :(x = real(z)),
+                 :(y = imag(z)),
+                 :(r = x + x),
+                 :(s = muladd(x, x, y*y)),
+                 as...,
+                 :(muladd($ai, z, $b)))
+    else
+        _evalpoly(z, p)
+    end
+end
+evalpoly(z::Complex, p::Tuple{<:Any}) = p[1]
+
+
+"""
+    evalpoly(z::Complex, p::Tuple)
+Evaluate the polynomial ``\\sum_k p[k] z^{k-1}`` for the coefficients `p[1]`, `p[2]`, ...;
+that is, the coefficients are given in ascending order by power of `z`. This function
+generates efficient code using a Goertzel-like algorithm specialized for complex arguments
+,*without* loops unrolled at compile time. Use this method when the number of coefficients
+is not known at compile time.
+The Goertzel-like algorthim is described in Knuth's Art of Computer Programming,
+Volume 2: Seminumerical Algorithms, Sec. 4.6.4.
+# Example
+```jldoctest
+julia> evalpoly(2 + im, (1, 2, 3))
+14 + 14im
+```
+"""
+evalpoly(z::Complex, p::AbstractVector) = _evalpoly(z, p)
+
+function _evalpoly(z::Complex, p)
+    length(p) == 1 && return p[1]
+    N = length(p)
+    a = p[end]
+    b = p[end-1]
+
+    x = real(z)
+    y = imag(z)
+    r = 2x
+    s = muladd(x, x, y*y)
+    for i in N-2:-1:1
+        ai = a
+        a = muladd(r, ai, b)
+        b = p[i] - s * ai
+    end
+    ai = a
+    muladd(ai, z, b)
+end
+
 """
     @horner(x, p...)
 
 Evaluate `p[1] + x * (p[2] + x * (....))`, i.e. a polynomial via Horner's rule.
 """
 macro horner(x, p...)
-    ex = esc(p[end])
-    for i = length(p)-1:-1:1
-        ex = :(muladd(t, $ex, $(esc(p[i]))))
-    end
-    ex = quote local r = $ex end # structure this to add exactly one line number node for the macro
-    return Expr(:block, :(local t = $(esc(x))), ex, :r)
+    esc(:(evalpoly($x, ($(p...),))))
```

Comment:
shouldn't the `esc` only be over the arguments `x` and `p`. (Same for below.)

Reply:
My understanding is that `esc`ing the entire expression is equivalent, but I could be wrong.

Reply:
I guess one difference is that if I `esc` the entire expression, that relies on the user having `evalpoly` in their workspace, so this could give unexpected results if the user defines their own `evalpoly`. I'll change it.

Reply:
Okay, I think I changed it correctly. I'll wait for the tests to pass before clicking resolve.